code/cdw.py

```diff
@@ -0,0 +1,259 @@
+import pandas as pd
+import nibabel as nib
+from itertools import compress
+
+
+def preprocess(bold, labels):
+    """
+    Minimal preprocessing of BOLD data
+    
+    Given a 4D BOLD-data detrends and z-scores it
+    run-wise ("chunks" in labels).
+
+    Parameters
+    ----------
+    bold : Nifti1Image
+        Original BOLD data.
+    labels : DataFrame
+        Labels for MRI frames.
+
+    Returns
+    -------
+    bold_prep : Nifti1Image
+        Preprocessed data.
+    """
+    
+    print('Copying data...')
+    prep = bold.get_fdata()
+    
+    print('Done. Begin preprocessing...')
+    for c in labels.chunks.unique():
+        print('Processing chunk ' + str(c))
+        inds = labels.query('chunks == @c').index
+        this_prep = prep[:,:,:,inds]
+        this_prep = signal.detrend(this_prep)
+        this_prep = stats.zscore(this_prep, axis=-1)
+        
+        # scipy z-score does not watch out for zero-division, so
+        # we have to correct those ourselves. Replacing nans with
+        # zeros makes sense as long as the input is all zeros, too.
+        if prep[:,:,:,inds][np.isnan(this_prep)].sum() == 0:
+            this_prep[np.isnan(this_prep)] = 0
+        else:
+            raise Exception('Unexpected nans! Aborting.')
+        
+        prep[:,:,:,inds] = this_prep
+    
+    bold_prep = nib.Nifti1Image(prep, bold.affine, bold.header)
+        
+    print('Done.')
+    return bold_prep
+
+def drop_by_labels(bold, labels, bad_labels):
+    """
+    Drop unwanted labels and fMRI frames
+    
+    Takes a 4D BOLD image and drops frames that correspond to unwanted
+    labels, and also drops those labels from the labels DataFrame.
+
+    Parameters
+    ----------
+    bold : Nifti1Image
+        Full 4D BOLD image.
+    labels : DataFrame
+        Full list of frame labels.
+    bad_labels : list of str
+        Unwanted labels.
+
+    Returns
+    -------
+    bold_subset : Nifti1Image
+        Selected BOLD frames.
+    labels_subset : DataFrame
+        Selected frame labels.
+
+    """
+    
+    good_inds = labels.query('labels not in @bad_labels').index
+    labels_subset = labels.loc[good_inds]
+    
+    print('Dropping unwanted frames...')
+    frames_subset = bold.get_fdata()[:,:,:,good_inds]
+    bold_subset = nib.Nifti1Image(frames_subset, bold.affine, bold.header)
+    
+    print('Done.')
+    return (bold_subset, labels_subset)
+    
+
+def reorder_by_labels(bold, labels):
+    """
+    Reorder labels and frames of 4D BOLD image
+    
+    Sorts labels (of frames) alphabetically and then reorders BOLD frames
+    to match that order.
+
+    Parameters
+    ----------
+    bold : Nifti1Image
+        Original 4D BOLD image.
+    labels : DataFrame
+        Original frame labels.
+
+    Returns
+    -------
+    bold_sorted : Nifti1Image
+        Sorted 4D BOLD image..
+    labels_sorted : DataFrame
+        Sorted frame labels.
+
+    """
+    
+    labels['fmri_i'] = range(len(labels))
```

Comment:
`fmri_i` is suboptimal as a name. For one, the variable is called `bold`. Furthermore, after the sorting, this is no longer the current index, it's the index in the original bold variable. How about `orig_bold_index` as a name?

Reply:
After careful thought, I agree. 🤔